Add Data.Bits to base

stefan-hoeck opened the pull request:
This PR adds a new interface for bitwise operations to base. This is part of #983. Implementations for `Bits64` and `Integer` will be added as soon as a new compiler including the changes from  #1026 is released.

I decided against adding an implementation for `Int`. The `Int` type seems not to be specified very well. It is, for instance, unclear, whether it is supposed to be a signed 64bit or 32bit-precision integer, or whether this depends on the backend and operating system. Also, some bitwise operations on negative numbers seem to behave rather strangely. For instance, on Chez Scheme (64bit Linux), `prim__shl_Int 1 63 = 0`, while on GHC Haskell, the same expression yields `-9223372036854775808`. I'm nut sure whether I should file a bug report about this, or whether we should just agree on not officially supporting `Int` via `Data.Bits` for now.

Finally: For the time being we only support `shiftL`, `shfitR`, `and`, `or`, and `xor` as primitive operations. All other bitwise operations are implemented in terms of these core functions. This is probably not ideal, since typically backends provide optimized versions of these functions already (especially `popCount` is most certainly far too slow). I'm not sure what the general opinion about adding additional primitives (which have to be supported across all backends) is. I'd suggest, we release this with unoptimized implementations for the time being, adding additional primitives on demand.

gallais reviewed (commented):
Why is it that `shift`, `bit`, and all other operations talking about
positions take an `a` rather than a `Nat`?

stefan-hoeck commented:
> Why is it that shift, bit, and all other operations talking about
> positions take an a rather than a Nat?

That's the types of the primitive functions, which I didn't touch. If you consider casting a Nat to a primitive to be low enough, we could opt for more safety here and use Nats for positioning. People requiring raw performance could still call the prim__ versions directly.

peterohanley commented:
Can shiftL/R be removed from the prelude, then? They're not used anywhere but libs/contrib/System/Random.idr and it's a bit odd to have duplicated members of an interface but for only one type just floating around. https://github.com/idris-lang/Idris2/blob/8102e9e49510d783b1e9d587e97676f20802995e/libs/prelude/Prelude/Interfaces.idr#L65

stefan-hoeck commented:
> Can shiftL/R be removed from the prelude, then? They're not used anywhere but libs/contrib/System/Random.idr and it's a bit odd to have duplicated members of an interface but for only one type just floating around.

Oh dear, I forgot about that one. As stated in the initial comment, I didn't plan to add an instance of `Data.Bits` for `Int` since `Int` is not well specified. How should we go about this? I could add an instance of `Bits` but not `FiniteBits`, since the bit size of `Int` can vary across backends. There is, for instance, a discussion going on, whether `Int` on JS backends should be represented by `Number` (see  #801), in which case it could not be considered a 64bit number on that backend.

stefan-hoeck commented:
Seems like `shiftL` and `shiftR` are also used in the compiler in `src/Core/Primitives.idr`. gallais, could you quickly comment on this: Will it be safe to remove `shiftL` and `shiftR` from the Prelude and replace them with calls to the `prim__` versions in `src/Core/Primitives.idr`?

Also, should we consider removing bit ops (the primitives as well) from `Int` altogether? I think the main issue here is, that we use modular arithmetic to deal with `Int` overflows, which leads to incorrect behavior when left shifting positive values in such a way that they should actually become negative. We could take the position that users should stick to Bits types (or Integer, where the bit ops are properly supported by the backends) if they need to do bit operations.

gallais commented:
> Seems like shiftL and shiftR are also used in the compiler in src/Core/Primitives.idr. gallais, could you quickly comment on this: Will it be safe to remove shiftL and shiftR from the Prelude and replace them with calls to the prim__ versions in src/Core/Primitives.idr?

Yes that would be fine given it's what we are doing for all of the other primitives.

> Also, should we consider removing bit ops (the primitives as well) from Int altogether?

I think we should fix `Int` to a specific width and provide `Int32` and `Int64` versions
of the primitives so that user may pick the precision they want. As we have already seen
with floats (#623) letting different backends use different precisions is a *very* bad
idea for consistency.

stefan-hoeck commented:
> We could maybe have an associated index type in the Bits interface?

Note, that this would not prevent us from shifting an 8 bit number by - for instance - 64 bit, since we could just call `shiftL` repeatedly. It would, however, make sure we stayed within bounds if one of the backends chose to throw an exception in case of a too large shift. I'm also a bit concerned about the additional verbosity for using such basic functions (a weak argument, I know). As we often shift by a fixed number of bits (known at compile time), we should at least have an implementation of `fromInteger` with some auto implicits for `Subset`, so users are still able to conveniently write something like

```idris
the Bits8 13 `shiftL` 2
```

> I think we should fix Int to a specific width and provide Int32 and Int64 versions of the primitives so that user may pick the precision they want.

This sounds very reasonable. We'll still have to decide what the precision of `Int` should be (probably 64 bit, as this is how it is implemented right now in JS and Scheme), but that discussion should probably go into #1048.

gallais commented:
> since we could just call shiftL repeatedly.

Right, it's not as important for `shift(L/R)` but it is a bit more
for `complementBit` and `bit` and `clearBit` and `setBit`.

> we should at least have an implementation of fromInteger with some auto implicits for Subset

We know that `lt` on `Nat` reflects `LT`. So we can indeed have a

```idris
fromInteger : (k : Nat) -> {n : Nat} -> {auto 0 _ : lt k n === True} -> Subset Nat (`LT` n)
```

that will be guaranteed to work.

stefan-hoeck commented:
Suggestions from reviewers have been implemented:

- use an `Index` type in Data.Bits
- added some proofs to Data.Nat (such as `ltReflectsLT`)
- remove `shiftL` and `shiftR` from Prelude and adjusted usage of these functions in the compiler and libraries

Thanks for all the valuable suggestions.

stefan-hoeck commented:
Argh, some typedd-book tests use `shiftR` from the Prelude. In order not to diverge too much from the book, I'll try to do the following over the weekend: Fix bitops implementations for `Int` and add a (hopefully correct) implementation of Data.Bits for `Int`, fixing it at 64bit as this is how it is implemented on all backends right now. I'll then fix the typedd tests.

stefan-hoeck commented:
This should once again be ready to review. The gambit test was failing, though I don't know why. I changed it to actually build an executable, which works fine on my system but takes ages on github, so I'm not sure what to do about this. In addition, I fixed the typedd tests to use Data.Bits and updated the docs accordingly.

stefan-hoeck commented:
Rebased and cleaned up commit history.

stefan-hoeck commented:
gallais, I think I need some help here. With a fresh make and install, tests like `chez/chez032` fail on my system. The culprit seems to be the following line

```idris
intmin : Int
intmin = prim__shl_Int 1 63
```

which is evaluated to 0 at compile time instead of -9223372036854775808. However, if I build and install the latest commit twice, everything works fine as it does in the CI on github. Do you have an idea what's going on here? If I change this to the following version, everything works OK:

```idris
shl63 : Int -> Int
shl63 = (`prim__shl_Int` 63)

intmin : Int
intmin = shl63 1
```

This seems also to be the reason why the gambit test fails.

stefan-hoeck commented:
Fixed and rebased (gambit problem was introduced by my own carelessness). Please review.

buzden commented on libs/base/Data/Bits.idr:
```diff
@@ -0,0 +1,192 @@
+
+  ||| Returns the bitwise complement of a value.
+  complement : a -> a
+
```
A tempting thing to ask is why not to have a function connecting a `bitSize` or `FiniteBits` with `Index` type in `Bits` like
```idris
0 bitToIndex : Fin bitSize -> Index
```
I can imagine why they might not be isomorphic, but the function working at least one way should exist.

stefan-hoeck replied:
Could you please elaborate on this? Do you see a use case for `bitToIndex` other than as a proof that `bitSize` and `Index` are properly correlated? Also, why might they not be isomorphic?

buzden replied:
> Do you see a use case for bitToIndex other than as a proof that bitSize and Index are properly correlated?

Well, no, what I originally suggested is purely a proof of proper correlation and that's why it has `0` quantity.

However, it can be useful for generic programming, i.e. for code working with any type satisfying `FiniteBits`. Now you cannot use any shifting and bit setting functions unless you know exactly the type. Having a `bitToIndex` function you can have generic functions (even if you'll need some requirement on `bitSize` value in its signature). But then, `bitsToIndex` shouldn't have zero quantity.

> Also, why might they not be isomorphic?

I can imagine a proper instance for finite-bits-type (like, e.g. `Bits32`) with `Index` being `Integer`. In this case there is no proper function of type `Index -> Fin bitSize`. This is very artificial example and maybe it's better to require an isomorphism between `Fin bitSize` and `Index` -- I'm just not sure about this.

When requiring an isomorphism, probably it's better to have a special data structure coding it (i.e. having functions in both directions and showing extensional equality of applications of these functions to each other with the identity function).

stefan-hoeck replied:
Thanks, buzden, for your kind explanations.

Since this PR is already quite big enough, I'd rather not blow it up any more by adding a new data type/interface for `Isomorphism`. But I'll give `bitsToIndex` a go, as it makes sense to correlate the two types.

gallais replied:
`Fin` is not efficient at the moment so you may want to have

```idris
bitToIndex : Subset Nat (`LT` bitSize) -> Index
```

instead.

stefan-hoeck replied:
Changed `bitToIndex` as suggested by gallais.

buzden commented on libs/base/Data/Bits.idr:
```diff
@@ -23,6 +24,13 @@ fromNat :  (k : Nat)
+||| Utility for implementing `bitsToIndex` in `FiniteBits`.
+export
+fromFin : Fin n -> Subset Nat (`LT` n)
+fromFin FZ     = Element 0 (LTESucc LTEZero)
```
I believe the right place for this function is in `Data.Fin`, somewhere near for `finToNat`, named like `finToNat'` or `finToBoundedNat`.

peterohanley commented on libs/base/Data/Bits.idr:
```diff
@@ -0,0 +1,200 @@
+  bitsToIndex : Fin bitSize -> Index {a}
+
+  ||| Returns the bitwise complement of a value.
+  complement : a -> a
```
Does `complement` belong in `FiniteBits`? `Integer` isn't `FiniteBits` yet

```
complement : Integer -> Integer
complement x = xor x -1
```

just like all the other implementations. Chez scheme `logxor` supports bignums at least. The missing piece is a `oneBits` in `Bits`

stefan-hoeck replied:
I assume with `oneBits` you mean the value with the first bit set to 1?

peterohanley replied:
No, all bits set to one. In twos complement that's -1, as in all the current implementations

peterohanley replied:
oneBits = complement zeroBits

stefan-hoeck replied:
I moved `complement` to Data.Bits and added `oneBits`. In addition, I moved `popCount` to `FiniteBits` as it makes more sense there. I just quickly checked with Haskell, where `popCount` is defined on `Bits`: There, `popCount ((-1) :: Integer)` returns -1, probably not what we want.

peterohanley replied:
👍

Just for completeness, complement in Bits prevents an instance for "unsigned bignums" e.g. Nat, but I don't think that instance is a good idea anyway, and adding a third class is not warranted here imo.

buzden commented on libs/base/Data/Fin.idr:
```diff
@@ -54,6 +55,13 @@ finToNatInjective FZ     (FS _) Refl impossible
+finToBoundedNat FZ     = Element 0 (LTESucc LTEZero)
+finToBoundedNat (FS x) = let Element k prf = finToBoundedNat x
+                          in Element (S k) (LTESucc prf)
+
```
Being rebased to the current `master`, the second case could be simplified to
```idris
finToBoundedNat (FS x) = bimap S LTESucc $ finToBoundedNat x
```

stefan-hoeck replied:
Neat! However, I'll rebase this and undo the addition of `finToBoundedNat`. After changing `bitsToIndex` to the more efficient type suggested by gallais, `finToBoundedNat` is unrelated to this PR.

gallais commented:
Ugh. Didn't realise this was taking the gambit test from ~15 minutes to 45+ minutes.
We should maybe consider turning that test off if it cannot be made to be faster.

stefan-hoeck commented:
> Ugh. Didn't realise this was taking the gambit test from ~15 minutes to 45+ minutes.
> We should maybe consider turning that test off if it cannot be made to be faster.

I find writing tests for the gambit and racket backends to be quite a pain, to be honest. And yet, these tests are important if we continue supporting these backends as implementations can vary slightly across different Scheme flavors. So yeah, feel free to turn off the gambit test. We probably still need to think about how to make sure, the gambit (and racket) backends behave correctly.
